Update types and handle data_base64 structured.

evankanderson opened the pull request:
- Add sane defaults for encoding
  - Unfortunately, defaults for structured and binary need to be *different*
   - Push types through interfaces
- Make it easy to call 'ToRequest' using Marshaller defaults
- Add tests for above

- Resolves #28  and slighly improves #26 

- I discovered that the types for `data_marshaller` and `data_unmarshaller` had to vary between structured and unstructured as a *user* of the SDK. In particular, it seemed that `UnmarshalJSON` wants to pass a `str` value to `data_unmarshaller`, while `UnmarshalBinary` wants to pass a `typing.IO`.

- I started by figuring out the necessary types for `data_marshaller` and `data_unmarshaller`. Along the way, I realized that binary CloudEvents data was not being handled correctly for JSON. See #28 and https://github.com/cloudevents/spec/blob/v1.0/json-format.md#31-handling-of-data. Annoyingly, if the data is not of type "binary", the structured format suggests passing it through without an extra layer of enveloping, which leads to some special cases in `MarshalJSON`.

- How to verify it

I added tests

- One line description for the changelog

```
Update python types and handle structured `data_base64`.
```


Note that this is probably **a breaking API change** because the expected types for `data_marshaller` and `data_unmarshaller` changed and all previous clients needed to supply these.

evankanderson commented:
I figured out how to fix the lint warnings; since it appears that the current configuration checks *both* W503 and W504, which are contradictory.

evankanderson commented:
@denismakogon any interest in this?

evankanderson commented:
Ping?

evankanderson commented:
Ping 2

On Fri, May 29, 2020 at 10:36 AM Evan Anderson <notifications@github.com>
wrote:

> Ping?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudevents/sdk-python/pull/34#issuecomment-636096525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XEN6U5QILJ6MNGANUPFTRT7XDHANCNFSM4M4QY57Q>
> .
>

grant commented:
@evankanderson Who are you pinging?

evankanderson commented:
I think @denismakogon is the only one who can approve/merge, but maybe @duglin can assist?

duglin commented:
I can merge but I'd need at least one person who knows python to review... I don't know python

duglin commented:
@di ^^^

grant commented:
Ping @evankanderson. It looks like you have a couple pending requested changes here.

evankanderson commented on cloudevents/sdk/event/base.py:
```diff
@@ -12,10 +12,16 @@
 import json
 import typing
 
+# Use consistent types for marshal and unmarshal functions across
+# both JSON and Binary format.
+MarshallerType = typing.Callable[[typing.Any], typing.Union[bytes, str]]
+UnmarshallerType = typing.Callable[[typing.IO], typing.Any]
```
I tried adding this to `sdk/__init__.py`. What do you think of that?

di replied:
I definitely prefer `cloudevents.types`, is there a reason not to use that?

evankanderson commented on cloudevents/sdk/event/base.py:
```diff
@@ -110,23 +116,41 @@ def Set(self, key: str, value: object):
-        return io.BytesIO(json.dumps(props).encode("utf-8"))
-
-    def UnmarshalJSON(self, b: typing.IO, data_unmarshaller: typing.Callable):
+        data = ""
+        if "data" in props:
+            data = data_marshaller(props.get("data"))
+            del props["data"]
```
Should we call `data_marshaller` even if the event doesn't have a data payload?

evankanderson replied:
Actually, looking at this further, I think we're including a `data` field in the JSON serialization when maybe we shouldn't (if there was no payload specified).

evankanderson commented on cloudevents/sdk/converters/structured.py:
```diff
@@ -35,16 +35,18 @@ def read(
-    ) -> (dict, typing.IO):
+        self,
+        event: event_base.BaseEvent,
+        data_marshaller: event_base.UnmarshallerType
```
Whoops, thanks!

evankanderson commented:
@di Thanks for the previous review; ready for another look.

di commented on cloudevents/sdk/__init__.py:
```diff
@@ -0,0 +1,16 @@
+#    under the License.
+
+# Convenience aliases
+from cloudevents.sdk.event.base import MarshallerType, UnmarshallerType  # noqa: F401
```
I think these should just be moved here instead (although I would prefer them to be in `cloudevents/types/__init__.py`)

evankanderson replied:
How about `cloudevents/sdk/types.py`? It feels strange to have the type definitions at a higher level than the rest of the SDK.

di replied:
I think the `sdk` submodule is actually the strange part... this whole library is the "sdk" 🙂 But that's probably a good compromise.

evankanderson replied:
I agree that it's strange to have the library under "sdk", but I don't want to rock the boat in this PR.

di replied:
I appreciate that! I agree we can make that a separate issue.

evankanderson commented:
I think I managed to appease the DCO bot, and got to polish up my `git` skills on the side.

di commented:
@evankanderson FYI, it looks like there are failing tests now.

evankanderson commented:
Yes, merging in #47 seems to have added some tests that I need to fix, looking.

evankanderson commented:
I've updated this PR to include the comments I made in #26 after I discovered conflicts during merge. Sorry that this made things a lot bigger.

I'm going to be on a camping trip all week, so feel free to crib from/adopt some of the changes in other PRs if they make sense and you want them to get in before Friday.

grant commented:
Thanks for the PR Evan.

> I'm going to be on a camping trip all week, so feel free to crib from/adopt some of the changes in other PRs if they make sense and you want them to get in before Friday.

Thanks. I'll see if we can fix the current checks and get this PR in shape to merge.

Checks:

```py

./cloudevents/sdk/types.py:22:36: S100 First argument on the same line
UnmarshallerType = typing.Optional[typing.Callable[[
                                   ^
./cloudevents/sdk/types.py:22:52: S100 First argument on the same line
UnmarshallerType = typing.Optional[typing.Callable[[
                                                   ^
./cloudevents/sdk/http_events.py:15:1: F401 'copy' imported but unused
import copy
^
./cloudevents/sdk/http_events.py:18:1: F401 'io' imported but unused
import io
^
./cloudevents/sdk/http_events.py:18:1: I100 Import statements are in the wrong order. 'import io' should be before 'import json'
import io
```

cumason123 commented:
Needs to update samples/http-cloudevents sample code

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
-            headers: dict = {},
-            data_unmarshaller: typing.Callable = lambda x: x,
+    @classmethod
+    def FromHttp(
+        cls,
+        data: typing.Union[str, bytes],
+        headers: dict = {},
+        data_unmarshaller: types.UnmarshallerType = None
     ):
```
Why do we decide to not follow the [pep8 convention for naming of python methods](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):
pep8 says they should be:
```python
    def from_http(...) -> ...:
        ...
```

Also can we make sure to do "complete type hints":
```suggestion
    def from_http(
        cls,
        data: typing.IO,
        headers: typing.Dict[str, typing.Any],
        data_unmarshaller: types.UnmarshallerType = None
    ) -> CloudEvent:
```

evankanderson replied:
Hmm -- it turns out that I can't use `-> CloudEvent` to document the return type, because the class is not defined until after all the functions have been defined (including this one).

Thoughts?

MarcBoissonneault replied:
> Hmm -- it turns out that I can't use `-> CloudEvent` to document the return type, because the class is not defined until after all the functions have been defined (including this one).
> 
> Thoughts?

for this add the following at the top:

MarcBoissonneault replied:
For this:
> typing.IO expects to have a read() method, but a quick survey of 3 popular Python libraries (flask, requests, sanic), suggests that the normal exposure of the request body is via a string where the library handles the read, rather than exposing the underlying socket.

I just used the exact same typehints that the current implementation is using here: https://github.com/cloudevents/sdk-python/blob/master/cloudevents/sdk/marshaller.py#L45

BUT after reading your comment, I think you are right!
And for the headers I also think you are right!

So, double 👍

evankanderson replied:
It turns out that [PEP 563](https://www.python.org/dev/peps/pep-0563/) is only supported in Python 3.7 or later.

MarcBoissonneault replied:
and that's exactly why the added this feature in the "future" module.

And this line in the imports:
```
from __future__ import annotations
```

And it should work.

evankanderson replied:
It doesn't work on Python 3.6, which is in the test suite.

I tried adding a PR to do that, and the 3.6 tests (only) failed. If we can abandon 3.6, I'd be happy to make the change.

di replied:
[Python 3.6 won't be EOL until Dec 2021](https://python-release-cycle.glitch.me/), so I think we should probably continue to support it.

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
+
+        return cls(attrs, event.data)
+
+    def __init__(self, attributes: dict = {}, data: typing.Any = None):
```
Can we make sure to do "complete type hints":
```suggestion
    def __init__(self, attributes: typing.Dict[str, typing.Any] = None, data: typing.Any = None):
```

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
+
+        return marshaller.NewDefaultHTTPMarshaller().ToRequest(event, format, data_marshaller)
+
+    # Attribute access is managed via Mapping type
```
❤️

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
+                "type": "cloudevent.event.type",
+                "specversion": "0.2"
             }
         :type headers: dict
```
```suggestion
        :type attributes: typing.Dict[str, typing.Any]
```

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
         """
-        converter_type = converters.TypeBinary if self.isbinary else \
-            converters.TypeStructured
+        marshaller_by_format = {converters.TypeStructured: lambda x: x,
+                                converters.TypeBinary: json.dumps}
```
this instantiation doesn't need to be re-created on every call, therefore this could be moved to a module variable.

MarcBoissonneault commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
+                                converters.TypeBinary: json.dumps}
+        if data_marshaller is None:
+            data_marshaller = marshaller_by_format[format]
+        obj_by_type = {"1.0": v1.Event, "0.3": v03.Event}
```
this instantiation doesn't need to be re-created on every call, therefore this could be moved to a module variable.

cumason123 commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,148 @@
+        if 'time' not in self._attributes:
+            self._attributes['time'] = datetime.datetime.now(
+                datetime.timezone.utc).isoformat()
+
```
In addition to that I guess in either this or a future PR we want to restore type checking on fields

MarcBoissonneault commented on tox.ini:
```diff
@@ -17,6 +17,7 @@ commands =
 [flake8]
-ignore = H405,H404,H403,H401,H306,S101,N802,N803,N806,I202,I201
+# See https://gitlab.com/pycqa/flake8/-/issues/466 for extend-ignore vs ignore
+extend-ignore = H405,H404,H403,H401,H306,S101,N802,N803,N806,I202,I201
```
All those "ignore" seems to indicate that we are not following standard python conventions ;)

evankanderson replied:
https://flake8.pycqa.org/en/latest/user/error-codes.html

Not sure what these are; I discovered that `extend-ignore` was having issues with W503/W504, where flake8 changed the behavior.

* H405, H404, H403, H401 are all about docstrings and spaces (I noticed some of these)
* S101 - Multi-line missing trailing comma
* N802, N803, N806 - function names lowercase, argument name lowercase, functiov in variable lowercase.
* I201, I202 - Newlines between import groups

Maybe we can clean these up in a separate PR before 1.0.0?

MarcBoissonneault replied:
FOR SURE we can change that in another PR!!! 
It was just a comment about this whole repo where there seems to have lot of ignored rules... 
Usually when we ignore rules it's because we don't like them or because we don't know the standard of the language.

Another thoughts here, personally I prefer `pylint` as linter (compare to `flake8`) since you can ignore rule by name instead of by "rule code" that are so hard (impossible) to remember.
For example, see this stackoverflow comment from 2014: https://stackoverflow.com/questions/4341746/how-do-i-disable-a-pylint-warning/23542817#23542817

evankanderson replied:
I totally agree on the code not feeling pythonic -- the existing code feels very Java-ish to me.

With that said, coming from go-land, I hate linters because it introduces human debug-fix loops where presumably the machine could do the formatting for me.

di replied:
I agree (these linting settings also predate me). I'd prefer if we just used https://pypi.org/project/black/, which includes an auto-formatter.

cumason123 commented:
Can we adjust the sample code at samples/http-cloudevents, or should we do that in a separate PR?

evankanderson commented:
> For the most part everything looks good. I do have a couple design questions/nits however.
> 
> To my understanding, a consumer must first strip cloudevent fields of the ce- prefix before instantiating a CloudEvent. Could we perhaps do this field stripping internally to reduce overhead for consumers?

I think your understanding is colored by looking at the API through the lens of HTTP-Binary, with an assumed JSON payload. 😁 Ideally, the CloudEvents class would be both transport and payload independent, so you could send an "image/jpeg" event as a POST reply, or store a CloudEvent as as an AVRO structured format in Kafka.

To instantiate a CloudEvent, the customer should provide the *attribute* names, rather than the HTTP-Binary *header* names. If you want to convert an HTTP-Binary request to a CloudEvent use `CloudEvent.FromHttp(body, headers)`, which should take an **un-decoded** body and possibly a `data_marshaller` if the payload is something exotic, rather than JSON.

> 
> ```python
> from flask import Flask, request
> from cloudevents.sdk.http_events import CloudEvent
> app = Flask(__name__)
> 
> @app.route("/", methods=["POST"])
> def echo():
>     attributes = {}
>     
>     contenttype = request.headers.get('Content-Type')
>     if contenttype is not None:
>         attributes['Content-Type'] = contenttype
>         
>     # unsure if passing None quietly fails thus I verify if field exists before
>     # passing it into the CloudEvent   
>     ce_type = request.headers.get('ce-type', request.form.get('type'))
>     if ce_type is not None:
>         # CloudEvent atm explicitly wants non ce- prefixed cloudevent fields
>         attributes['type'] = ce_type
>         
>     ce_id = request.headers.get('ce-id', request.form.get('id'))
>     if ce_id is not None:
>         attributes['id'] = ce_id
>         
>     ce_source = request.headers.get('ce-source', request.form.get('source'))
>     if ce_source is not None:
>         attributes['source'] = ce_source
>        
>     ce_specversion = request.headers.get('ce-specversion', request.form.get('specversion'))
>     if ce_specversion is not None:
>         attributes['specversion'] = ce_specversion
>     
>     data = request.form.get('data_base64', request.form.get('data'))
>     event = CloudEvent(attributes, data)
>     
>     # Do some work
>     return '', 204
> 
> if __name__ == '__main__':
>     app.run()
>     
> ```
> 
> However this is how a consumer would ideally like to instantiate events:
> 
> ```python
> from flask import Flask, request
> from cloudevents.sdk.http_events import CloudEvent
> app = Flask(__name__)
> 
> @app.route("/", methods=["POST"])
> def home():
>     headers = dict(request.form)
>     body = dict(request.headers)
>     
>     # CloudEvent automatically parses the http headers/body
>     event = CloudEvent(headers, body)
>     
>     # Do some work 
>     return '', 204
> 
> if __name__ == '__main__':
>     app.run()
>     
> ```

I'd suggest the following for reading CloudEvents, which will work whether or not the CloudEvents payload is JSON:

```python
from flask import Flask, request
from cloudevents.sdk.http_events import CloudEvent
app = Flask(__name__)

@app.route("/", methods=["POST"])
def home():
    # CloudEvent automatically parses the http headers/body, whether or not the body is JSON
    event = CloudEvent.FromHttp(request.get_data(), request.headers)

    # Do some work 
    return '', 204

if __name__ == '__main__':
    app.run()
    
```

evankanderson commented on cloudevents/tests/test_http_events.py:
```diff
@@ -110,13 +113,13 @@ def test_emit_binary_event(specversion):
+        "Content-Type": "text/plain"
     }
-    event = CloudEvent(test_data, headers=headers)
+    data = json.dumps(test_data)
```
I thought that we were testing the sender side in `test_binary_to_request` on line 236. My concern here was that there could be a reciprocal bug in sending and receiving (for example, if the receive did not strip the `ce-` prefix from attributes, and then the response encoding did not add `ce-`, it would be impossible to tell if we use the library on both paths).

cumason123 replied:
Fair point. We can leave this as it is and in a future PR I'll just add an additional test verifying  cloudevents on both ends work just to ensure the sdk can talk to itself.

evankanderson replied:
I actually ended up adding a loop through the SDK in `test_roundtrip_non_json_event`.

evankanderson commented on cloudevents/tests/test_http_events.py:
```diff
@@ -125,17 +128,18 @@ def test_emit_binary_event(specversion):
         if key != 'Content-Type':
-            assert r.headers[key] == headers[key]
+            attributeKey = key[3:]
+            assert r.headers[attributeKey] == headers[key]
```
This was deliberate -- I'm not replying with a CloudEvent, but with HTTP headers with the same names as the CloudEvents attributes. Would you prefer to have the attribute map at the beginning and then convert to HTTP-Binary headers afterwards, rather than stripping the `ce-` prefix here?

I'm willing to do an encode/decode loop through the library, but I'm concerned that in doing so we might end up accidentally producing something which can talk to itself but not to other implementations.

evankanderson commented on cloudevents/tests/test_http_events.py:
```diff
@@ -144,11 +148,10 @@ def test_emit_structured_event(specversion):
         "specversion": specversion,
         "data": test_data
     }
-    event = CloudEvent(body, headers=headers)
```
See the argument above -- I don't think we want to round-trip the CloudEvent on both sides because it makes it harder to test the protocol conformance (though the code would happily be much simpler).

evankanderson commented:
> Needs to update samples/http-cloudevents sample code

Any thoughts on whether/how we could add a test that the sample works?

evankanderson commented:
Hope everyone had a good 4th of July weekend!

I've updated the README.md; please take a look at the new API(s), since I've made breaking changes. I suspect that's okay given the pre-1.0 library and generally low adoption. If there are further changes we should make, it's probably better to make them sooner than later.

I've added a test for uploading bz2-encoded binary data and round-tripping that through the library. I didn't do a raw test for that because the content is opaque, but I can add one if it feels important.

evankanderson commented on samples/http-cloudevents/client.py:
```diff
@@ -14,57 +14,56 @@
 import sys
 import io
 from cloudevents.sdk.http_events import CloudEvent
+from cloudevents.sdk.types import converters
```
Whoops, thanks!

evankanderson commented on cloudevents/sdk/http_events.py:
```diff
@@ -11,183 +11,157 @@
+            data_unmarshaller = _json_or_string
+
+        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
+            v1.Event(), headers, data, data_unmarshaller)
```
Until we've parsed the event, it's hard to find the specversion. Fortunately, since we extract everything from the Event and throw it away, all we need are the parsing, which seems to be the same path for v3 and v1.

I agree that it's not ideal, but I think it will take a bunch of refactoring to separate the attribute extraction from the representation.

cumason123 replied:
Alright. I'll make an issue for this once the PR gets merged.

evankanderson commented on samples/http-cloudevents/server.py:
```diff
@@ -20,11 +20,10 @@
+    event = CloudEvent.from_http(request.get_data(), request.headers)
 
     # print the received CloudEvent
     print(f"Received CloudEvent {event}")
```
What's even stranger is:

```python
>>> e = CloudEvent.from_http('test', {'ce-specversion': '1.0', 'ce-id': '123', 'ce-source': 'ttt', 'ce-type': 'yyy'})
>>> e
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __repr__ returned non-string (type bytes)
>>> e.__repr__()
b'{"specversion": "1.0", "id": "123", "source": "ttt", "type": "yyy", "time": "2020-07-06T20:27:17.132295+00:00", "extensions": {}, "data": "test"}'
```

I added a `.decode()` to the bytes representation from the structured body.

cumason123 replied:
Yeah, I tried e.__repr__() and it worked. I was so convinced that something on my end was bugged that I didn't even mention it lol.

grant commented:
Are we good to merge this @evankanderson @cumason123?

cumason123 commented:
Possible easy refactor:

```python
...
from cloudevents.sdk import converters

headers, body = event.to_http(converters.TypeBinary)
```

could be refactored into:

```python
headers, body = event.to_http(isbinary=True)
```

User no longer has to import an additional path. Functionally converters only has two possible types atm anyways.

evankanderson commented:
I'd prefer to keep the types for now.

My suspicion is that `to_http` and `from_http` are actually not the correct methods on a CloudEvent, because then we'll need to add `to_kafka`, `to_amqp`, etc.

I agree that it would be nice to have a better way to type `event.to_http(converters.TypeBinary)`, but I think I might prefer to spell it `http.binary.emit(event)` or `http.structured.parse(headers, body)`.

evankanderson commented:
It looks like there are approvals and no requested changes. @cumason123 -- are you comfortable with this being committed?
